### PR TITLE
feat(optimizers): SinkGD width transfer, spectral norm, MD sphere + fused Triton kernels

### PR DESCRIPTION
**Pull request by winglian**
Builds on #3763. Adds two experimentally-validated optimizer axes to SinkGD plus an opt-in fused Triton path, all default-off — the default (all-flags-off) update is byte-identical to the shipped optimizer.

## Feature A: width transfer + spectral normalization

Width sweeps (tiny random-init Llama, d_model 256→2048, LR grids, coordinate checks) established that SinkGD's raw update is **Adam-class** (`η ∝ 1/d_in`), and that pinning the update's operator norm to `sqrt(d_out/d_in)` via power iteration converts it to Muon-class — which both transfers LR across width (flat optimum, lr≈2 at every width) and trains to 6–21% lower loss than the `1/d_in` rule at every width tested. It also bounds the deep-layer activation blow-up plain SinkGD shows at depth (peak residual RMS ~92 at 16 layers → ~18).

- `sinkgd_base_width` (+ `sinkgd_lr_width_exponent`): width-aware `alpha_eff = sinkgd_lr_scale * (base_width/d_in)**exp`. Unset → old scalar behavior.
- `sinkgd_spectral_norm` (+ `sinkgd_spectral_norm_iters`, `sinkgd_spectral_target: unit|muon`): warm-started power-iteration operator-norm rescale (~2 matvecs/layer, one O(d_in) state vector, +1–3% step time; ~free on B200).
- `sinkgd_base_width` + `spectral_target: muon` are two width corrections and double-count — rejected at construction.
- FSDP2: on rows-sharded weights the power iteration adds one small vector all-reduce per iteration over `dp_shard` (same shape/group as the existing Sinkhorn norm reduce); the matrix is never gathered; state round-trips through checkpoints.

## `sinkgd_md_sphere` (experimental): magnitude–direction weight sphere

The A+B combine from the study: each 2D weight held on its enable-time Frobenius sphere, SR-Sinkhorn direction update spectral-normalized to unit operator norm, analytic sphere projection (no learnable gains — SinkGD's row/col balancing makes them redundant, measured). Tightest deep-layer activation bound of any variant and wins at depth (16-layer loss 1.60 vs 1.73 spectral-only), but its optimal LR is width-dependent (`lr_opt ∝ d_model**-0.6`) with a narrow optimum → experimental, off by default, plain spectral norm is the recommended path. `SinkGDMD`/`DistSinkGDMD` subclasses; FSDP2 adds one per-matrix scalar all-reduce for the sphere norm.

## `sinkgd_fused_kernel` (opt-in): fused Triton path

One kernel per SR-Sinkhorn iteration replaces the compiled loop's ~4 passes (the column scale folds forward into the next kernel, the power-iteration vectors, and the apply; the MD projection uses the analytic Frobenius form). Auto tall/wide grid selection keeps both full matrices and heavily-sharded wide-short local shards occupied. Same all-reduce count as the compiled dist path. Falls back to the compiled path for cols-sharded (TP) dims, `bf16_stochastic_round`, missing Triton, and small single-device spec/md tensors (`sinkgd_fused_min_numel`). Numerics equivalent at bf16 rounding scale (scales carried fp32 end-to-end) but not byte-identical → off by default.

Optimizer-step speedups vs the compiled path (production classes, 7-matrix layer):

| | single GPU | 2-rank dist | 8-rank-equiv shards |
|---|---|---|---|
| B200 (sm_100) | 1.4–1.75× | 1.0–1.9× | 1.06–1.6× |
| H100 (sm_90) | 1.1–1.3× | **1.2–2.2×** | 1.4–1.8× |

Every mode × size × card cell measured ≥ ~1× after the size gate (H100 exposed a small-matrix spec/md epilogue regression that the larger-L2 cards hid; gated).

## Also in this PR

Two perf rounds on the compiled spectral/MD paths (bf16 tensor-core matvecs, final-column-scale folding, analytic sphere projection): isolated step overhead spectral +31%→+13%, MD +40%→+22%.

## Validation

- 36 unit + 12 FSDP2 (torchrun 2-rank) + builder + e2e tests; suites pass on sm_120 (RTX PRO 6000), sm_100 (2×B200), and sm_90 (2×H100).
- Fused-vs-compiled parity tests for all modes on square/wide/3D-expert shapes, single-device and DTensor-sharded, tall and wide kernel regimes; sphere invariance; fallback gating.
- Full experiment write-ups (phase results, LR-transfer sweeps, norm-growth studies, kernel regime maps, all bench tables) are in the study logs; happy to attach any table in more detail.

Papers: [SinkGD arXiv:2502.06742](https://arxiv.org/abs/2502.06742), [HP-transfer arXiv:2512.05620](https://arxiv.org/abs/2512.05620), [MD decoupling arXiv:2606.25971](https://arxiv.org/abs/2606.25971).

## Summary by CodeRabbit

* **New Features**
  * Added new SinkGD configuration options for spectral normalization, width-aware scaling, fused Triton execution, and an experimental sphere-based variant.
  * Expanded distributed optimizer support so these modes work across sharded and multi-GPU training setups.

* **Documentation**
  * Documented the new SinkGD options, including valid combinations and fallback behavior.

* **Tests**
  * Added unit and end-to-end coverage for the new SinkGD modes, configuration validation, and distributed/fused behavior.

### CODE REVIEW

File: docs/optimizers.qmd [hunk 1]
```diff
@@ -150,6 +150,89 @@ optim_args:
   sinkgd_lr_scale: 0.05  # α scale applied to linear-layer updates
 ```
 
+#### Width transfer (`sinkgd_base_width`) and spectral normalization
+
+Two optional, default-off knobs help the learning rate transfer across model width and
+condition the update. They are **mutually exclusive** — both correct for width, so enabling
+both double-counts (validated and rejected at construction).
```
[the remark this answers is not in the document]
Comment (winglian) [reply docs/optimizers.qmd]: Fixed in 72fbaf6ab — reworded to scope the mutual exclusion to `sinkgd_base_width` + `sinkgd_spectral_target: muon`, and noted that the default `unit` target composes with `base_width`.

File: src/axolotl/utils/optimizers/sinkgd_triton.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    B, m_local, n = g.shape
+    m_glob = m_global if m_global is not None else m_local
+    sqrt_n, sqrt_m = float(n) ** 0.5, float(m_glob) ** 0.5
+    dev = g.device
+
+    x_buf = torch.empty_like(g)
+    colsq_a = torch.full((B, n), float(m_glob), device=dev, dtype=torch.float32)
+    colsq_b = torch.empty(B, n, device=dev, dtype=torch.float32)
+    nsplits, cols_per = _plan_splits(B, m_local, n, block_m, block_n)
+    grid = (triton.cdiv(m_local, block_m), nsplits, B)
```
[the remark this answers is not in the document]
Comment (winglian) [reply src/axolotl/utils/optimizers/sinkgd_triton.py:426]: Addressed in 72fbaf6ab, with one deviation from the suggestion: calling `p.contiguous()` before dispatch would be wrong here — the `_apply` kernel updates `p` in place, so the update would land in the copy and be silently dropped. Instead, `_fused_ok` now gates on `p.is_contiguous()` (non-contiguous params fall back to the compiled path, which handles arbitrary strides via `p.copy_`), and `fused_sinkgd_step` raises on a non-contiguous `p` so external callers fail loudly instead of corrupting storage. Verified: non-contiguous view param takes the compiled path and matches the flag-off result exactly with the rest of the storage untouched; contiguous params still dispatch fused.

File: tests/e2e/multigpu/test_sinkgd_fsdp2.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        lr=1.0,
+        sinkgd_lr_scale=1.0,
+        sinkgd_spectral_norm=True,
+        sinkgd_spectral_norm_iters=sn_iters,
+        sinkgd_spectral_target=target_mode,
+        process_group=mesh["dp_shard"].get_group(),
+    )
+    opt2.load_state_dict(sd)
+    torch.testing.assert_close(opt2.state[w]["specnorm_u"], u)
+
```
[the remark this answers is not in the document]
Comment (winglian) [reply tests/e2e/multigpu/test_sinkgd_fsdp2.py:127]: Fixed in 72fbaf6ab — both tests now broadcast the RNG-generated reference tensors from rank 0 before sharding (and before computing `tn0` in the MD test), matching `test_dist_fused_matches_compiled`. Re-ran the 2-rank suite: 12 passed.
